verification/sign.go

```diff
@@ -0,0 +1,189 @@
+	for i := range label {
+		label[i] = byte(i)
+	}
+
+	tbs := make([]byte, digestLen)
+	for i := range tbs {
+		tbs[i] = byte(i)
+	}
+
+	signedData, err := c.Sign(handle, label, SignFlags(IsSymmetric), tbs)
+	if err != nil {
+		t.Fatalf("[FATAL]: Error while signing %v", err)
+	}
+
+	// Rerun with same label and compare signature emitted.
+	signedDataWithSameLabel, err := c.Sign(handle, label, SignFlags(IsSymmetric), tbs)
+	if err != nil {
+		t.Fatalf("[FATAL]: Error while signing %v", err)
+	}
+
+	// It is sufficient to check signatureR, signatureS is all zeroes
+	if !bytes.Equal(signedDataWithSameLabel.HmacOrSignatureR, signedData.HmacOrSignatureR) {
+		t.Errorf("[ERROR]: Signature varies for same label, want %v but got %v", signedData.HmacOrSignatureR, signedDataWithSameLabel.HmacOrSignatureR)
+	}
+
+	// Rerun with different label, signature must change this time
+	newLabel := make([]byte, digestLen)
+	for i := range newLabel {
+		newLabel[i] = byte(0)
+	}
+
+	signedDataWithDiffLabel, err := c.Sign(handle, newLabel, SignFlags(IsSymmetric), tbs)
+	if err != nil {
+		t.Fatalf("[FATAL]: Error while signing %v", err)
+	}
+
+	if bytes.Equal(signedDataWithDiffLabel.HmacOrSignatureR, signedData.HmacOrSignatureR) {
+		t.Errorf("[ERROR]: Signature must vary for different label despite having same toBeSigned content, want new signature but got old %v", signedData.HmacOrSignatureR)
+	}
+}
+
+// Checks whether the digital signature returned by Sign command can be verified
+// using public key in signing key certificate returned by CertifyKey command.
+// Inspite of the DPE profile supporting symmetric key, for symmetric signing it must be enabled
+// explicitly in Sign command. Else asymmetric signing is used as default.
+func testAsymmetricSigning(d TestDPEInstance, c DPEClient, t *testing.T, handle *ContextHandle, digestLen int) {
+	var err error
+	flags := SignFlags(0)
+
+	seqLabel := make([]byte, digestLen)
+	for i := range seqLabel {
+		seqLabel[i] = byte(i)
+	}
+
+	tbs := make([]byte, digestLen)
+	for i := range tbs {
+		tbs[i] = byte(i)
+	}
+
+	signResp, err := c.Sign(handle, seqLabel, flags, tbs)
+	if err != nil {
+		t.Fatalf("[FATAL]: Error while signing %v", err)
+	}
+
+	// Get signing key certificate using CertifyKey command
+	certifiedKey, err := c.CertifyKey(handle, seqLabel, CertifyKeyX509, CertifyKeyFlags(0))
+	if err != nil {
+		t.Fatalf("[FATAL]: Could not CertifyKey: %v", err)
+	}
+
+	// Check certificate structure
+	var leafCert *x509.Certificate
+	if leafCert, err = x509.ParseCertificate(certifiedKey.Certificate); err != nil {
+		t.Fatalf("[FATAL]: Could not parse certificate using crypto/x509: %v", err)
+	}
+
+	// Read public key
+	var ec elliptic.Curve
+	x := new(big.Int).SetBytes(certifiedKey.Pub.X)
+	y := new(big.Int).SetBytes(certifiedKey.Pub.Y)
+
+	if digestLen == 32 {
+		ec = elliptic.P256()
+	} else if digestLen == 48 {
+		ec = elliptic.P384()
+	}
+
+	publicKey := ecdsa.PublicKey{Curve: ec, X: x, Y: y}
```

Comment:
Can you add a check that `publicKey` and `leafCert.PublicKey` are the same key? (Would also be fine if you'd rather make this check in the CertifyKey test instead)

Reply:
Have added this in CertifyKey.go, but yet to raise as PR.
I will be raising as a separate PR, as I have considerable amount of changes in this file. 
Adding in this PR will again increase the changes and make review difficult.

I have tested in that file as below
```
// Checks CertifyKey command response against public key extracted from certificate returned in response
func checkCertifyKeyResponse(t *testing.T, x509Cert *x509.Certificate, response CertifiedKey, hashAlg asn1.ObjectIdentifier) {
	var err error

	publicKeyDer, err := x509.MarshalPKIXPublicKey(x509Cert.PublicKey)
	if err != nil {
		t.Fatalf("[FATAL]: Could not marshal pub key: %v", err)
	}

	// Parse the DER-encoded public key
	pubKey, err := x509.ParsePKIXPublicKey(publicKeyDer)
	if err != nil {
		t.Fatalf("[FATAL]: Failed to parse DER-encoded public key: %v", err)
	}

	if _, ok := pubKey.(*ecdsa.PublicKey); !ok {
		t.Fatal("[FATAL]: Public key is not a ecdsa key")
	}

	var pubKeyInResponse ecdsa.PublicKey

	if hashAlg.Equal(OidSHA384) {
		pubKeyInResponse = ecdsa.PublicKey{
			Curve: elliptic.P384(),
			X:     new(big.Int).SetBytes(response.Pub.X),
			Y:     new(big.Int).SetBytes(response.Pub.Y),
		}
	} else if hashAlg.Equal(OidSHA256) {
		pubKeyInResponse = ecdsa.PublicKey{
			Curve: elliptic.P256(),
			X:     new(big.Int).SetBytes(response.Pub.X),
			Y:     new(big.Int).SetBytes(response.Pub.Y),
		}
	} else {
		t.Errorf("[ERROR]: Unsupported hash algorithm.")
		return
	}

	if !(pubKeyInResponse.Equal(pubKey)) {
		t.Errorf("[ERROR]: Public key returned in response must match the Public Key Info in the leaf certificate.")
	}
}

```